tsl/src/remote/txn.c

```diff
@@ -754,7 +754,9 @@ remote_txn_persistent_record_write(TSConnectionId cid)
 
 	rel = table_open(catalog->tables[REMOTE_TXN].id, RowExclusiveLock);
 	persistent_record_insert_relation(rel, id);
-	table_close(rel, RowExclusiveLock);
 
+	/* Keep the table lock until transaction completes in order to
+	 * synchronize with distributed restore point creation */
+	table_close(rel, NoLock);
```

Comment:
Does it mean that all distributed transactions will be serialized due to keeping this lock until the end? cc @pmwkaa @erimatnor

Reply:
To my understanding this lock should not affect other distributed transaction accessing this table, because this lock would block only on conflicting key updates, which should not ever happen. We are locking this table with higher priority lock in create distributed restore point function, so basically it will lock other sessions untill that function completes

Reply:
Isn't it, every transaction calls this function insert a record into it? The lock locks entire the remote transaction table, isn't it?
Since the function is called before sending the prepare to data nodes, the lock will be kept while negotiating prepare until the commit is sent.

Reply:
Not its not, to my understanding this lock does not lock other transactions, unless they have conflicting keys. Otherwise we would also use it in the create_distributed_restore_point function, attempt to use it was unsuccessful

Reply:
To my understanding `rel = table_open(catalog->tables[REMOTE_TXN].id, RowExclusiveLock);` locks entire table. No keys to specific transactions are used here.

Reply:
Again, it's not. Row-level locks do not affect data querying, they block only writers to the same row.

Reply:
Actually it is a [table level lock](https://www.postgresql.org/docs/12/explicit-locking.html#LOCKING-TABLES). However it doesn't conflict with itself, so I guess it conflicts only on the same row. It wasn't obvious to me, so you are right that it should not be an issue. Thank you for the patience.